src/node_buffer.cc

```diff
@@ -602,6 +602,80 @@ void Compare(const FunctionCallbackInfo<Value> &args) {
 }
 
 
+void IndexOfString(const FunctionCallbackInfo<Value>& args) {
+  ASSERT(args[0]->IsObject());
+  ASSERT(args[1]->IsString());
+
+  ARGS_THIS(args[0].As<Object>());
+  node::Utf8Value str(args.GetIsolate(), args[1]);
+  uint32_t offset = args[2]->Uint32Value();
+
+  if (str.length() == 0) {
+    return args.GetReturnValue().Set(
+        MIN(static_cast<uint32_t>(obj_length), offset));
+  }
+  if (str.length() + offset > obj_length || obj_length == 0) {
+    return args.GetReturnValue().Set(-1);
+  }
+
+  char* result = strstr(obj_data + offset, *str);
+  args.GetReturnValue().Set(
+      result ? static_cast<int32_t>(result - obj_data) : -1);
+}
+
+
+void IndexOfBuffer(const FunctionCallbackInfo<Value>& args) {
+  ASSERT(args[0]->IsObject());
+  ASSERT(args[1]->IsObject());
+
+  ARGS_THIS(args[0].As<Object>());
+  Local<Object> buf = args[1].As<Object>();
+  uint32_t offset = args[2]->Uint32Value();
+  size_t buf_length = buf->GetIndexedPropertiesExternalArrayDataLength();
+  char* buf_data =
+      static_cast<char*>(buf->GetIndexedPropertiesExternalArrayData());
+  if (buf_length > 0)
+    CHECK_NE(buf_data, nullptr);
+
+  if (buf_length + offset > obj_length || obj_length == 0)
+    return args.GetReturnValue().Set(-1);
+
+  int32_t ret = -1;
+  uint32_t i = offset;
+
+  for (; i < obj_length; i++) {
+    if (obj_data[i] == buf_data[0]) {
+      ret = memcmp(obj_data + i, buf_data, buf_length);
+      if (ret == 0)
+        break;
+    }
+  }
+
+  args.GetReturnValue().Set(ret == 0 ? i : -1);
+}
+
+
+void IndexOfNumber(const FunctionCallbackInfo<Value>& args) {
+  ASSERT(args[0]->IsObject());
+  ASSERT(args[1]->IsNumber());
+
+  ARGS_THIS(args[0].As<Object>());
+  int32_t val = args[1]->Uint32Value();
+  uint32_t offset = args[2]->Uint32Value();
+
+  if (offset > obj_length || obj_length == 0)
+    return args.GetReturnValue().Set(-1);
+
+  char data[2];
+  data[0] = val;
```

Comment:
We're not going to throw. I was also considering the search for byte sequence. e.g. `b.indexOf(0xbada55)`, and think it is a good solution. Though users will need to know that byte sequences over `0x20000000000000` loose precision. Also, I'm not worried about endianness. The input data is endiannes independent, and it's up to the user to know what the data looks like that they're searching for.

@bnoordhuis What would be the fastest way to convert a number to a byte sequence that can be searched for?